# Aceleradev Ciencia de Dados
**Autor: Erick Amorim**

## Imputação de Dados

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("dados/train.csv")

In [3]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

**Definindo um objeto auxiliar para verificar as variáveis (colunas), tipos de dados e valores faltantes**

In [4]:
aux = pd.DataFrame({"colunas": df.columns, "tipos": df.dtypes, "percentual_faltantes": (df.isna().sum()/df.shape[0])*100 })

In [5]:
aux

colunas    tipos  percentual_faltantes
PassengerId  PassengerId    int64              0.000000
Survived        Survived    int64              0.000000
Pclass            Pclass    int64              0.000000
Name                Name   object              0.000000
Sex                  Sex   object              0.000000
Age                  Age  float64             19.865320
SibSp              SibSp    int64              0.000000
Parch              Parch    int64              0.000000
Ticket            Ticket   object              0.000000
Fare                Fare  float64              0.000000
Cabin              Cabin   object             77.104377
Embarked        Embarked   object              0.224467

**Substituindo pela Média ou Mediana em variável numérica**

In [6]:
df["Age"] = df["Age"].fillna(df["Age"].mean())

#df["Age"].fillna(df["Age"].median())

**Verificando o percentual de faltantes de 'Age' após a substituição**

In [7]:
aux = pd.DataFrame({"colunas": df.columns, "tipos": df.dtypes, "percentual_faltantes": (df.isna().sum()/df.shape[0])*100 })

In [8]:
aux

colunas    tipos  percentual_faltantes
PassengerId  PassengerId    int64              0.000000
Survived        Survived    int64              0.000000
Pclass            Pclass    int64              0.000000
Name                Name   object              0.000000
Sex                  Sex   object              0.000000
Age                  Age  float64              0.000000
SibSp              SibSp    int64              0.000000
Parch              Parch    int64              0.000000
Ticket            Ticket   object              0.000000
Fare                Fare  float64              0.000000
Cabin              Cabin   object             77.104377
Embarked        Embarked   object              0.224467

**Substituição em variável categórica**

veja que 77.10% dos valores em 'Cabin' estão faltando.

In [9]:
df["Cabin"] = df["Cabin"].fillna("Unknown")

**Contagem dos valores em 'Cabin'**

In [10]:
df["Cabin"].value_counts()

Unknown        687
G6               4
B96 B98          4
C23 C25 C27      4
F33              3
              ... 
A16              1
B102             1
T                1
B39              1
D10 D12          1
Name: Cabin, Length: 148, dtype: int64

*Número de linhas em df*

In [11]:
df.shape[0]

891